lxd/cluster/heartbeat.go

```diff
@@ -66,18 +146,33 @@ func Heartbeat(gateway *Gateway, cluster *db.Cluster) (task.Func, task.Schedule)
 				return err
 			}
 
+			offlineThreshold, err = tx.NodeOfflineThreshold()
+			if err != nil {
+				return err
+			}
+
 			return nil
 		})
 		if err != nil {
 			logger.Warnf("Failed to get current cluster nodes: %v", err)
 			return
 		}
 
+		// If this leader node hasn't sent a heartbeat recently, then its node state records
+		// are likely out of date, this can happen when a node becomes a leader.
+		// Detect this and just sent a raft-only partial state list for the first round.
```

Comment:
I don't understand this part, we operate on data coming from the global database, therefore `allNodes` here contains every single node in the cluster, so I'm not sure how we could possibly be getting a partial node list.

Reply:
@stgraber I will update the comment to better explain this, but this is how I understand how the heartbeat works, and why I felt the need to add this:

- With the exception of new nodes joining the cluster, heartbeats are normally leader initiated and read the last heartbeat time of each node out of dqlite before sending that onto every other node.
-  The last heartbeat time for each node is compared against the time now and if the difference is greater than the offline threshold setting (defaults 20s) then the node is marked as offline in the heartbeat info that is sent to each node.
- When the nodes receive the heartbeat info containing a list of all the other nodes, when refreshing the forkdns peer list, it excluded any "offline" nodes so as not to delay DNS lookups against apparently offline nodes.
- When each node responds positively to a heartbeat, the leader then updates dqlite database with the new heartbeat time of that node.

This all works fine, except when a leader leaves and a new leader needs to be elected, or when the cluster loses quorum and there is no leader.

When this happens, for a period of time (and in my tests when killing a leader and waiting for a new one be elected this is over 20s) there is no leader, and so no leader initiated heartbeats are sent. 

This causes the last heartbeat time in dqlite db to become stale for all nodes, even the ones that are up and running fine.

When a new node is elected and becomes leader, it begins sending heartbeats again. 

The problem is that first heartbeat it sends after it becomes leader it reads the stale node data our of dqlite (with last heartbeat times for all nodes being stale) and then checks those stale times against the time now. The duration is then often above the offline threshold and so the nodes are marked as "offline" and sent to all other nodes in the first heartbeat.

When all the other nodes receive the heartbeat, they see that all other nodes are offline and proceed to remove all of the forkdns peers in active use, causing a complete cluster DNS outage until the next heartbeat is sent with the updated heartbeat times from this heartbeat.

So I added a "last leader initiated heartbeat time" variable to each node, such that it can record the last time it was leader and whether it has recently sent a heartbeat. If it hasn't, then it considers the all node data to be partial (partial, because the node heartbeat times are stale) and proceeds to indicate this to all the nodes so that they don't trigger a forkdns peer refresh.

Note: when moving to using the offline threshold setting to define a node being offline, I had to calculate this on the leader as non-db nodes do not have access to this setting as it is stored in dqlite db. As such the `LastHeartbeat` in the request data is just there right now for informational purposes.

Whilst this approach does prevent the cluster wide DNS outage at each leader change, there is a potential for it to introduce unnecessary individual node DNS outages when the LXD process itself is down (perhaps being upgraded), as when LXD missing heartbeats beyond the "offline threshold" it will be considered "down", even if forkdns process is still running on that node.

Perhaps a nice addition in the future would be for the leader to initiate a dns query against each node to as part of the heartbeat and provide that in the heartbeat request data.

Perhaps this can be better worded to avoid the word "partial" or to clarify that "partial" in this case doesn't just mean the nodes themselves, but also their last check freshness.

Reply:
FWIW 20 seconds for the leader election to happens is a bit worrying. The election timeout is set to 3 seconds, I'd expect a new leader to be elected within 4/5 seconds. Maybe that's something we want to investigate.

In theory we should have a reasonably fast leader change (less 5 seconds) and a high enough offline threshold (20 seconds) that the problem described by @tomponline does not happen. But I'm mainly speculating, since I didn't try the whole thing.

Reply:
@freeekanayaka @stgraber this is the reason I originally picked 3*heartbeat interval for the classification of "forkdns peer offline", to give enough time for the cluster to settle after an "event" to avoid removing all DNS peers. I'll do some more tests this morning to get a firmer set of figures.

Reply:
@freeekanayaka here is an example:

4 node cluster.

Leader node runs and sends successful heartbeat round to all nodes finishing at 09:17.24 , then at 09:17:32 the leader is cleanly shutdown.

```
DBUG[07-11|09:17:19] Starting heartbeat round 
DBUG[07-11|09:17:19] Heartbeat updating local raft nodes to [{ID:1 Address:192.168.1.201:8443} {ID:2 Address:192.168.1.202:8443} {ID:3 Address:192.168.1.203:8443}] 
DBUG[07-11|09:17:19] Sending heartbeat to 192.168.1.201:8443 
DBUG[07-11|09:17:19] Sending heartbeat request to 192.168.1.201:8443 
DBUG[07-11|09:17:19] Successful heartbeat for 192.168.1.201:8443 
DBUG[07-11|09:17:22] Sending heartbeat to 192.168.1.203:8443 
DBUG[07-11|09:17:22] Sending heartbeat request to 192.168.1.203:8443 
DBUG[07-11|09:17:22] Successful heartbeat for 192.168.1.203:8443 
DBUG[07-11|09:17:24] Sending heartbeat to 192.168.1.204:8443 
DBUG[07-11|09:17:24] Sending heartbeat request to 192.168.1.204:8443 
DBUG[07-11|09:17:24] Successful heartbeat for 192.168.1.204:8443 
DBUG[07-11|09:17:24] Completed heartbeat round 
^CINFO[07-11|09:17:32] Received 'interrupt signal', exiting 
```

The node that eventually becomes leader's logs from that point onwards:

```
WARN[07-11|09:17:33] Error during dqlite proxy copy: tls: use of closed connection                                                          
DBUG[07-11|09:17:33] Database error: failed to begin transaction: failed to receive response: failed to receive header: failed to receive h$
ader: EOF                                                                                                                                   
WARN[07-11|09:17:33] Failed to get current cluster nodes: failed to begin transaction: failed to receive response: failed to receive header$
 failed to receive header: EOF                                                                                                              
DBUG[07-11|09:17:33] Disconnected event listener: 7f525314-3c60-497e-a6b4-9de8643710d1                                                      
DBUG[07-11|09:17:33] Found cert                               name=0                                                                        
DBUG[07-11|09:17:33] Found cert                               name=0                                                                        
DBUG[07-11|09:17:34] Found cert                               name=0                                                                        
DBUG[07-11|09:17:34] Found cert                               name=0                                                                        
DBUG[07-11|09:17:34] Dqlite: closing client                                                                                                 
DBUG[07-11|09:17:34] Found cert                               name=0                                                                        
DBUG[07-11|09:17:34] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
201:8443 attempt=0                                                                                                                          
DBUG[07-11|09:17:34] Found cert                               name=0                                                                        
DBUG[07-11|09:17:34] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/$
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=0                                      
DBUG[07-11|09:17:34] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
203:8443 attempt=0                                                                                                                          
DBUG[07-11|09:17:34] Dqlite: connection failed err=no available dqlite leader server found attempt=0                                        
DBUG[07-11|09:17:34] Found cert                               name=0                                                                        
DBUG[07-11|09:17:34] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
201:8443 attempt=1                                                                                                                          
DBUG[07-11|09:17:34] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/$
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=1                                      
DBUG[07-11|09:17:34] Found cert                               name=0                                                                        
DBUG[07-11|09:17:34] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
203:8443 attempt=1                                                                                                                          
DBUG[07-11|09:17:34] Dqlite: connection failed err=no available dqlite leader server found attempt=1                                        
DBUG[07-11|09:17:34] Found cert                               name=0                                                                        
DBUG[07-11|09:17:34] Found cert                               name=0                                                                        
DBUG[07-11|09:17:34] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
201:8443 attempt=2                                                                                                                          
DBUG[07-11|09:17:34] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/$
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=2                                      
DBUG[07-11|09:17:34] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
203:8443 attempt=2                                                                                                                          
DBUG[07-11|09:17:34] Dqlite: connection failed err=no available dqlite leader server found attempt=2                                        
DBUG[07-11|09:17:35] Found cert                               name=0                                                                        
DBUG[07-11|09:17:35] Found cert                               name=0                                                                        
DBUG[07-11|09:17:35] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
201:8443 attempt=3                                                                                                                          
DBUG[07-11|09:17:35] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/$
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=3                                      
DBUG[07-11|09:17:35] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
203:8443 attempt=3                                                                                                                          
DBUG[07-11|09:17:35] Dqlite: connection failed err=no available dqlite leader server found attempt=3                                        
DBUG[07-11|09:17:35] Found cert                               name=0                                                                        
DBUG[07-11|09:17:35] Found cert                               name=0                                                                        
DBUG[07-11|09:17:36] Found cert                               name=0                                                                        
DBUG[07-11|09:17:36] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
201:8443 attempt=4                                                                                                                          
DBUG[07-11|09:17:36] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/d
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=4
DBUG[07-11|09:17:36] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
203:8443 attempt=4
DBUG[07-11|09:17:36] Dqlite: connection failed err=no available dqlite leader server found attempt=4
DBUG[07-11|09:17:36] Found cert                               name=0
WARN[07-11|09:17:36] Error during dqlite proxy copy: tls: use of closed connection
DBUG[07-11|09:17:37] Found cert                               name=0
DBUG[07-11|09:17:37] Found cert                               name=0
DBUG[07-11|09:17:37] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
201:8443 attempt=5
DBUG[07-11|09:17:37] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/d
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=5
DBUG[07-11|09:17:37] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
203:8443 attempt=5
DBUG[07-11|09:17:37] Dqlite: connection failed err=no available dqlite leader server found attempt=5
DBUG[07-11|09:17:37] Found cert                               name=0
DBUG[07-11|09:17:38] Found cert                               name=0
DBUG[07-11|09:17:38] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
201:8443 attempt=6                                                                                                                          
DBUG[07-11|09:17:38] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/d
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=6                                      
DBUG[07-11|09:17:38] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
203:8443 attempt=6
DBUG[07-11|09:17:38] Dqlite: connection failed err=no available dqlite leader server found attempt=6
DBUG[07-11|09:17:38] Found cert                               name=0
DBUG[07-11|09:17:39] Found cert                               name=0
DBUG[07-11|09:17:39] Found cert                               name=0
DBUG[07-11|09:17:39] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
201:8443 attempt=7
DBUG[07-11|09:17:39] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/$
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=7
DBUG[07-11|09:17:39] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
203:8443 attempt=7
DBUG[07-11|09:17:39] Dqlite: connection failed err=no available dqlite leader server found attempt=7
DBUG[07-11|09:17:39] Found cert                               name=0
DBUG[07-11|09:17:40] Found cert                               name=0
DBUG[07-11|09:17:40] Found cert                               name=0
DBUG[07-11|09:17:40] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
201:8443 attempt=8
DBUG[07-11|09:17:40] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/$
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=8
DBUG[07-11|09:17:40] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
203:8443 attempt=8
DBUG[07-11|09:17:40] Dqlite: connection failed err=no available dqlite leader server found attempt=8
DBUG[07-11|09:17:40] Found cert                               name=0
DBUG[07-11|09:17:41] Found cert                               name=0
DBUG[07-11|09:17:41] Found cert                               name=0
DBUG[07-11|09:17:41] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
201:8443 attempt=9
DBUG[07-11|09:17:41] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/$
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=9
DBUG[07-11|09:17:41] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
203:8443 attempt=9
DBUG[07-11|09:17:41] Dqlite: connection failed err=no available dqlite leader server found attempt=9
DBUG[07-11|09:17:41] Found cert                               name=0
DBUG[07-11|09:17:42] Found cert                               name=0
DBUG[07-11|09:17:42] Found cert                               name=0
DBUG[07-11|09:17:42] Found cert                               name=0
DBUG[07-11|09:17:42] Found cert                               name=0
DBUG[07-11|09:17:42] Found cert                               name=0
DBUG[07-11|09:17:42] Dqlite: connected address=192.168.1.201:8443 attempt=10
DBUG[07-11|09:17:42] Connecting to a remote LXD over HTTPs
WARN[07-11|09:17:42] Failed to get events from node 192.168.1.202:8443: Unable to connect to: 192.168.1.202:8443
DBUG[07-11|09:17:43] Connecting to a remote LXD over HTTPs
WARN[07-11|09:17:43] Failed to get events from node 192.168.1.202:8443: Unable to connect to: 192.168.1.202:8443
DBUG[07-11|09:17:44] Starting heartbeat round
DBUG[07-11|09:17:44] Heartbeat updating local raft nodes to [{ID:1 Address:192.168.1.201:8443} {ID:2 Address:192.168.1.202:8443} {ID:3 Addr$
ss:192.168.1.203:8443}]
WARN[07-11|09:17:44] Sending partial heartbeat as node list out of date since: 0001-01-01 00:00:00 +0000 UTC
DBUG[07-11|09:17:45] Sending heartbeat to 192.168.1.203:8443
DBUG[07-11|09:17:45] Sending heartbeat request to 192.168.1.203:8443
DBUG[07-11|09:17:45] Successful heartbeat for 192.168.1.203:8443
DBUG[07-11|09:17:48] Sending heartbeat to 192.168.1.202:8443
DBUG[07-11|09:17:48] Sending heartbeat request to 192.168.1.202:8443
DBUG[07-11|09:17:48] Failed heartbeat for 192.168.1.202:8443: failed to send HTTP request: Put https://192.168.1.202:8443/internal/database$
 dial tcp 192.168.1.202:8443: connect: connection refused
DBUG[07-11|09:17:49] Sending heartbeat to 192.168.1.204:8443
DBUG[07-11|09:17:49] Sending heartbeat request to 192.168.1.204:8443
DBUG[07-11|09:17:49] Successful heartbeat for 192.168.1.204:8443
DBUG[07-11|09:17:49] Completed heartbeat round
```

This took 11s until the point of starting to send the next heartbeat (I've seen this take between 10 and 18s in my tests), however the time from the last heartbeat time of the nodes being written to dqlite (remember the last heartbeat round finished at 09:17:24) is 20s.

In this case, with an offline threshold of 20s, its easy to see how the last heartbeat time of all nodes could be >=20s (meaning we would consider all nodes to be offline) and take down the cluster DNS for 10s.

And technically the last heartbeat time is recorded as the same time for all of the nodes at the end of the heartbeat run, whereas if we wanted it to be accurate then it should be recorded as the time the heartbeat responded. Because of the way the heartbeats are skewed, this could be up to 7s out. But this would just exacerbate the problem above.

Reply:
Here's an example if it taking 18s to become leader after old leader was killed.

```
WARN[07-11|09:01:41] Error during dqlite proxy copy: tls: use of closed connection                                                          
DBUG[07-11|09:01:41] Database error: failed to begin transaction: failed to receive response: failed to receive header: failed to receive h$
ader: EOF                                                                                                                                   
WARN[07-11|09:01:41] Failed to get current cluster nodes: failed to begin transaction: failed to receive response: failed to receive header$
 failed to receive header: EOF                                                                                                              
DBUG[07-11|09:01:41] Disconnected event listener: 77b00817-7e46-4f08-8700-175f7099467a                                                      
DBUG[07-11|09:01:42] Found cert                               name=0                                                                        
DBUG[07-11|09:01:42] Dqlite: closing client                                                                                                 
DBUG[07-11|09:01:42] Found cert                               name=0                                                                        
DBUG[07-11|09:01:42] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
201:8443 attempt=0                                                                                                                          
DBUG[07-11|09:01:42] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/$
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=0                                      
DBUG[07-11|09:01:42] Found cert                               name=0                                                                        
DBUG[07-11|09:01:42] Found cert                               name=0                                                                        
DBUG[07-11|09:01:42] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
203:8443 attempt=0                                                                                                                          
DBUG[07-11|09:01:42] Dqlite: connection failed err=no available dqlite leader server found attempt=0                                        
DBUG[07-11|09:01:42] Found cert                               name=0                                                                        
DBUG[07-11|09:01:42] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1$
201:8443 attempt=1                                                                                                                          
DBUG[07-11|09:01:42] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/d
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=1
DBUG[07-11|09:01:42] Found cert                               name=0
DBUG[07-11|09:01:42] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
203:8443 attempt=1
DBUG[07-11|09:01:42] Dqlite: connection failed err=no available dqlite leader server found attempt=1
DBUG[07-11|09:01:42] Found cert                               name=0
DBUG[07-11|09:01:42] Found cert                               name=0
DBUG[07-11|09:01:42] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
201:8443 attempt=2
DBUG[07-11|09:01:42] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/d
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=2
DBUG[07-11|09:01:42] Found cert                               name=0
DBUG[07-11|09:01:42] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
203:8443 attempt=2
DBUG[07-11|09:01:42] Dqlite: connection failed err=no available dqlite leader server found attempt=2
DBUG[07-11|09:01:43] Found cert                               name=0
DBUG[07-11|09:01:43] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
201:8443 attempt=3
DBUG[07-11|09:01:43] Found cert                               name=0
DBUG[07-11|09:01:43] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/d
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=3
DBUG[07-11|09:01:43] Found cert                               name=0
DBUG[07-11|09:01:43] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
203:8443 attempt=3
DBUG[07-11|09:01:43] Dqlite: connection failed err=no available dqlite leader server found attempt=3
DBUG[07-11|09:01:43] Found cert                               name=0
DBUG[07-11|09:01:44] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
201:8443 attempt=4
DBUG[07-11|09:01:44] Found cert                               name=0
DBUG[07-11|09:01:44] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/d
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=4
DBUG[07-11|09:01:44] Found cert                               name=0
DBUG[07-11|09:01:44] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
203:8443 attempt=4
DBUG[07-11|09:01:44] Dqlite: connection failed err=no available dqlite leader server found attempt=4
DBUG[07-11|09:01:45] Found cert                               name=0
DBUG[07-11|09:01:45] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
201:8443 attempt=5
DBUG[07-11|09:01:45] Found cert                               name=0
DBUG[07-11|09:01:45] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/d
atabase: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=5
DBUG[07-11|09:01:45] Found cert                               name=0
DBUG[07-11|09:01:45] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.
203:8443 attempt=5
DBUG[07-11|09:01:45] Dqlite: connection failed err=no available dqlite leader server found attempt=5
WARN[07-11|09:01:45] Error during dqlite proxy copy: tls: use of closed connection
DBUG[07-11|09:01:46] Found cert                               name=0                                                                 [0/477]
DBUG[07-11|09:01:46] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.203:8443 attempt=6
DBUG[07-11|09:01:46] Dqlite: connection failed err=no available dqlite leader server found attempt=6
DBUG[07-11|09:01:47] Found cert                               name=0
DBUG[07-11|09:01:47] Found cert                               name=0
DBUG[07-11|09:01:47] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.201:8443 attempt=7
DBUG[07-11|09:01:47] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/database: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=7
DBUG[07-11|09:01:47] Found cert                               name=0
DBUG[07-11|09:01:47] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.203:8443 attempt=7
DBUG[07-11|09:01:47] Dqlite: connection failed err=no available dqlite leader server found attempt=7
DBUG[07-11|09:01:48] Found cert                               name=0
DBUG[07-11|09:01:48] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.201:8443 attempt=8
DBUG[07-11|09:01:48] Found cert                               name=0
DBUG[07-11|09:01:48] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/database: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=8
DBUG[07-11|09:01:48] Found cert                               name=0
DBUG[07-11|09:01:48] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.203:8443 attempt=8
DBUG[07-11|09:01:48] Dqlite: connection failed err=no available dqlite leader server found attempt=8
DBUG[07-11|09:01:49] Found cert                               name=0
DBUG[07-11|09:01:49] Dqlite: server connection failed err=failed to establish network connection: 503 Service Unavailable address=192.168.1.201:8443 attempt=9
DBUG[07-11|09:01:49] Found cert                               name=0
DBUG[07-11|09:01:49] Dqlite: server connection failed err=failed to establish network connection: Head https://192.168.1.202:8443/internal/database: dial tcp 192.168.1.202:8443: connect: connection refused address=192.168.1.202:8443 attempt=9
DBUG[07-11|09:01:49] Found cert                               name=0
DBUG[07-11|09:01:49] Found cert                               name=0
DBUG[07-11|09:01:49] Found cert                               name=0
DBUG[07-11|09:01:49] Dqlite: connected address=192.168.1.203:8443 attempt=9
DBUG[07-11|09:01:49] Connecting to a remote LXD over HTTPs
WARN[07-11|09:01:49] Failed to get events from node 192.168.1.202:8443: Unable to connect to: 192.168.1.202:8443
DBUG[07-11|09:01:50] Found cert                               name=0
DBUG[07-11|09:01:50] Found cert                               name=0
DBUG[07-11|09:01:50] Connecting to a remote LXD over HTTPs
WARN[07-11|09:01:51] Failed to get events from node 192.168.1.202:8443: Unable to connect to: 192.168.1.202:8443
DBUG[07-11|09:01:52] Connecting to a remote LXD over HTTPs
WARN[07-11|09:01:52] Failed to get events from node 192.168.1.202:8443: Unable to connect to: 192.168.1.202:8443
DBUG[07-11|09:01:53] Connecting to a remote LXD over HTTPs
WARN[07-11|09:01:53] Failed to get events from node 192.168.1.202:8443: Unable to connect to: 192.168.1.202:8443
DBUG[07-11|09:01:54] Connecting to a remote LXD over HTTPs
WARN[07-11|09:01:54] Failed to get events from node 192.168.1.202:8443: Unable to connect to: 192.168.1.202:8443
DBUG[07-11|09:01:55] Connecting to a remote LXD over HTTPs
WARN[07-11|09:01:55] Failed to get events from node 192.168.1.202:8443: Unable to connect to: 192.168.1.202:8443
DBUG[07-11|09:01:56] Connecting to a remote LXD over HTTPs
WARN[07-11|09:01:56] Failed to get events from node 192.168.1.202:8443: Unable to connect to: 192.168.1.202:8443
DBUG[07-11|09:01:57] Connecting to a remote LXD over HTTPs
WARN[07-11|09:01:57] Failed to get events from node 192.168.1.202:8443: Unable to connect to: 192.168.1.202:8443
DBUG[07-11|09:01:58] Starting heartbeat round
DBUG[07-11|09:01:58] Heartbeat updating local raft nodes to [{ID:1 Address:192.168.1.201:8443} {ID:2 Address:192.168.1.202:8443} {ID:3 Address:192.168.1.203:8443}]
DBUG[07-11|09:01:58] Connecting to a remote LXD over HTTPs
WARN[07-11|09:01:58] Failed to get events from node 192.168.1.202:8443: Unable to connect to: 192.168.1.202:8443
WARN[07-11|09:01:58] Sending partial heartbeat as node list out of date since: 0001-01-01 00:00:00 +0000 UTC
```

Reply:
@stgraber @freeekanayaka as discussed I've re-worked the cluster hearbeats to send an initial stale heartbeat to all nodes marked as partial when a  node becomes leader. Then immediately after that run a normal heartbeat.